### DISCUSSION TITLE
Worker::waitPayload virtually disables custom posix signal handling

### DISCUSSION THREAD

**Discussion by l-x** [roadrunner-server/roadrunner#1678] [Q&A] [answered]
At the moment the `\Spiral\RoadRunner\Worker::waitPayload` is a blocking stream operation which virtually disables asynchronous signal handling in my worker scripts.

We have the absolute necessity to maintain a TCP connection to an internal service using manual heartbeat requests, regardless of whether the worker loop is currently waiting for tasks or not. We currently achieve this via `pcntl_alarm` callbacks, but this does not seem to be possible in Roadrunner at the moment.

Is there any way to call custom code at regular (configurable) intervals while `Worker::waitPayload` is waiting for data frames?

**Comment by rustatian** [comment]
Hey @l-x 👋🏻 
Might be it would be better to write a small custom plugin for RR and build it with Velox?

**Reply by l-x** [reply]
I'm not sure how we would achieve that by a rr plugin. Each worker has to maintain it's own tcp connection. Just to clarify what's that all about:

Some components require the use of [php-ampqlib](https://github.com/php-amqplib/php-amqplib) and the connections to the RabbitMQ server requires us to use heartbeats. It might happen that the http workers have nothing to do for some time and therefore are not able to communicate over this connection. After 60 seconds the RabbitMQ server closes the connection to the worker due to not having received the heartbeats from the client, which results in a crash of the worker as soon as a http request arrives and the php-ampqlib connection is used.

php-ampqlib ships with a component `PCNTLHeartbeatSender` to send heartbeat requests on a configured interval utilising `pcntl_alarm` and `pcntl_async_signals`. When using blocking stream operations (i. E. `fread` in `\Spiral\Goridge\StreamRelay::waitFrame`) no signal handlers are called while waiting for data which then leads to a immediate crash of the worker as soon as the blocking call returns.

**Reply by rustatian** [reply]
Just out of curiosity, what is the reason for not using RR's AMQP driver?

**Reply by l-x** [reply]
The main reason for not using RR's amqp driver is the lack of support for amqp transactions (https://www.rabbitmq.com/amqp-0-9-1-reference.html#tx.select) and the RabbitMQ RPC pattern (https://www.rabbitmq.com/tutorials/tutorial-six-php.html).

By setting `STDIN` to non-blocking (`\stream_set_blocking(STDIN, false);`) and changing

```php
$header = @\fread($this->in, 12);
```
to
```php
while (true) {
    $header = @\fread($this->in, 12);

    if ($header !== '') {
        break;
    }
    \usleep(50_000);
}
```
in `\Spiral\Goridge\StreamRelay::waitFrame` I am able to get the heartbeat sender going.

I guess I'm fine to use a custom relay class, but maybe I'm not the only one that needs non-blocking wait for job- and http-workers.

**Reply by rustatian** [reply]
Yeah, you're right, we don't have those things in the RR worker/RR AMQP driver yet, but that's pretty easy to add. RPC and transactions looks like a good feature request (and would be cool if you send us a ticket, thus I would be able to start working on that).
As for the PHP part (and since I'm not a PHP developer), I guess we could call @roxblnfk / @wolfy-j to answer your question.

**Reply by l-x** [reply]
That sounds like a good idea 😃

I guess I will send a ticket on Wednesday!

**Reply by roxblnfk** [reply]
Eah as you can see all the classes are extendable and have an interface. And it's OK to have some custom implementation.
We have [a task](https://github.com/roadrunner-php/goridge/issues/12) to make connections non blocking with fibers support but we have no plans about it.

**Reply by rustatian** [reply]
@wolfy-j  / @roxblnfk WDYT guys, do we have the capacity to implement/plan this feature in future?

**Reply by wolfy-j** [reply]
Hi, while we have a ticket we do not plan to take on this feature any time soon (maybe never). It is possible to make non blocking workers - like this example for RR1 - https://github.com/phpinnacle/amridge 

But the scope of work is too large while it does not provide any substantial benefits for most of the users.

I can only advice you:
1) Make your own worker by forking ours. If your PR might work without any BC we might even merge it back.
2) Describe the core issue outside of the scope of the worker, maybe some TCP connection pool on RR end will be better for resource sharing and healthchecks.

**Reply by l-x** [reply]
Just to clarify, my problem is not the Worker blocking my application but the relay blocking the whole PHP VM while waiting for data. I am able to solve this with a slightly modified StreamRelay.

**Reply by l-x** [reply, marked as answer]
To whom it may concern, we solved the issue with the following simple wrapper for the PHP StreamRelay:

```php
<?php

declare(strict_types=1);

namespace Derp;

use Spiral\Goridge\Frame;
use Spiral\Goridge\RelayInterface;
use Spiral\Goridge\StreamRelay;

use function stream_select;

class StreamRelayWrapper implements RelayInterface
{
    /**
     * @param resource $readStream
     * @param StreamRelay $relay
     */
    public function __construct(
        private $readStream,
        private readonly StreamRelay $relay,
    ) {
    }

    private function waitForData(): void
    {
        $null = null;

        while (true) {
            $readStreams = [$this->readStream];
            
            if (
                @stream_select(
                    $readStreams,
                    $null,
                    $null,
                    0,
                    200_000
                )
            ) {
                break;
            }
        }
    }

    public function waitFrame(): Frame
    {
        $this->waitForData();

        return $this->relay->waitFrame();
    }

    public function send(Frame $frame): void
    {
        $this->relay->send($frame);
    }

    public function hasFrame(): bool
    {
        return $this->relay->hasFrame();
    }
}

$relay = new StreamRelayWrapper(STDIN, new StreamRelay(STDIN, STDOUT));
```

**Reply by rustatian** [reply]
I think you won't need to use this wrapper anymore when I implement these additional features for the AMQP plugin, am I right?

**Reply by l-x** [reply]
Yeah, that's true :D
